i#3805 msvc2019: add support for Microsoft Visual Studio 2019 and Tools for Visual Studio 2019

StephanvanSchaik opened the pull request:
This addresses issue #3805 as follows:

- Remove old detection code for CMake versions older than 2.6.3 to detect `cl_path`. This was hindering CMake from finding the correct `cl_path`.
- Ensure `cl_path` is defined before `cpp2asm_support.cmake` gets included, such that `ml.exe` and `ml64.exe` for x86 and x64 respectively are found in the correct path.
- Specify `cl_path` as a hint when trying to find `lib.exe` and `dumpbin.exe`.
- The Windows SDK that is being used may be installed into `C:\Program Files (x86)\Windows Kits\10`, of which the installation path is covered by a different registry key, namely `HKEY_LOCAL_MACHINE\Software\Microsoft\Windows Kits\Installed Roots\KitsRoot10`. Check if the registry key points to an existing path, and use it as `kits_dir` if it does and then use that to derive the correct `sdk_bindir`. Otherwise fall back to the old method of detecting `sdk_bin` and `sdk_bindir`.
- When MFC is installed, this will try building drstats. Since drstats has hardcoded paths to `Uuid.Lib`, we now simply use `kits_dir` to specify the correct path to `Uuid.Lib`.

I have tested this with Tools for Visual Studio 2019 as follows:

```
git clone https://github.com/StephanvanSchaik/dynamorio -b i3805-msvc2019
cd dynamorio
mkdir build
cd build
cmake ..
cmd.exe
call "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Auxiliary\Build\vcvarsall.bat" amd64
msbuild.exe DynamoRIO.sln
```

This builds everything successfully. However, do note that you should not have Windows 11 SDK (version 10.0.22000.0) installed. Trying to build DynamoRIO against the Windows 11 SDK will fail with rc.exe emitting error RC2188 due to line 253 in `C:\Program Files (x86)\Windows Kits\10\Include\10.0.22000.0\um\winnt.h`. See also: https://bugs.python.org/issue45220 (EDIT: the fix for this is to ensure `_AMD64_` or `_X86_` are defined when building).

derekbruening commented:
Thank you for working on this -- this will be needed for #5195

derekbruening commented:
Adding a GA CI job that uses VS2019 might be the best way to have a test for this as part of this PR.  To focus on building (probably some tests will be flaky on the change of platform; could try running them first: maybe we'll get lucky and they'll be green) and separate test success to #5195, add `build_only` to the runsuite_wrapper args.

I.e., please add a windows-2019 job (probably need 2, one for 32-bit and one for 64-bit).  This could be inside the same file `.github/workflows/ci-windows.yml`: copy the existing ones and add `build_only`.

StephanvanSchaik commented:
@derekbruening You are welcome!

* I have updated my branch with upstream.
* I have added a workflow for vs2019-builds that builds successfully now and that currently passes the `build_only` argument when running `runsuite_wrapper`.
* I have fixed an issue where either `_AMD64_` or `_X86_` need to be defined on Microsoft Windows when using the Windows SDK 10.0.22000.x but weren't because either the generator did not define them (e.g. the Ninja generator has this issue), or because CMake did not pass the definitions to rc.exe (for the Microsoft Visual Studio generators).

derekbruening commented on .github/workflows/ci-windows.yml:
```diff
@@ -244,3 +244,68 @@ jobs:
         from: Github Action CI
+
+  ###########################################################################
+  # 32-bit and 64-bit VS2019 release builds:
```
Looks like this is release-build-only so it's skipping a bunch of debug code which might contain warnings or issues in vs2019.  Removing the `nontest_only` below should make it build both; or maybe better to replace `nontest_only` with `debug_only` to cover the most code in one job (else this job might become the slowest one) (we'll get release build covered with the switch in #5195).

derekbruening replied:
nit: It still says "release" here

derekbruening replied:
I edited it myself.

StephanvanSchaik commented on ext/drsyms/CMakeLists.txt:
```diff
@@ -70,7 +70,7 @@ set(DynamoRIO_USE_LIBC ON)
   # XXX: if we add any more of these .lib files we should share this code
   # (currently we have make/ntdll_imports.cmake and here).
-  find_program(LIB_EXECUTABLE lib.exe DOC "path to lib.exe")
+  find_program(LIB_EXECUTABLE lib.exe HINTS ${cl_path} DOC "path to lib.exe")
```
Good catch.

derekbruening commented:
> StephanvanSchaik force-pushed the i3805-msvc2019 branch from 9120d63 to d570f1d 15 hours ago

For the future, please don't force-push: it breaks the viewing of recent changes (important for reviewing), breaks the comment history, etc.  This is documented on multiple pages such as https://dynamorio.org/page_contributing.html

derekbruening commented:
add to allowlist

derekbruening commented:
run arm tests

StephanvanSchaik commented:
> > StephanvanSchaik force-pushed the i3805-msvc2019 branch from [9120d63](https://github.com/DynamoRIO/dynamorio/commit/9120d63aca45a9ee281dfd0ddfa923abcee971b4) to [d570f1d](https://github.com/DynamoRIO/dynamorio/commit/d570f1d2c350fa54f8e0d8eac0672da08c50cee7) 15 hours ago
> 
> For the future, please don't force-push: it breaks the viewing of recent changes (important for reviewing), breaks the comment history, etc. This is documented on multiple pages such as https://dynamorio.org/page_contributing.html

@derekbruening Sorry about that, I am a bit used to the Linux kernel reviewing process where you just e-mail patch sets, and essentially just re-open the pull requests for later versions to keep the commits clean. One of the last projects I worked with on Github instead prefered having the same PR open. If instead you prefer additional commits to address the reviews, then I am happy to do that too.

Also sorry about missing it in the contribution page, I did skim through that page to get an overall idea.

derekbruening commented:
> > > StephanvanSchaik force-pushed the i3805-msvc2019 branch from [9120d63](https://github.com/DynamoRIO/dynamorio/commit/9120d63aca45a9ee281dfd0ddfa923abcee971b4) to [d570f1d](https://github.com/DynamoRIO/dynamorio/commit/d570f1d2c350fa54f8e0d8eac0672da08c50cee7) 15 hours ago
> > 
> > 
> > For the future, please don't force-push: it breaks the viewing of recent changes (important for reviewing), breaks the comment history, etc. This is documented on multiple pages such as https://dynamorio.org/page_contributing.html
> 
> @derekbruening Sorry about that, I am a bit used to the Linux kernel reviewing process where you just e-mail patch sets, and essentially just re-open the pull requests for later versions to keep the commits clean. One of the last projects I worked with on Github instead prefered having the same PR open. If instead you prefer additional commits to address the reviews, then I am happy to do that too.
> 
> Also sorry about missing it in the contribution page, I did skim through that page to get an overall idea.

No problem, different systems can be confusing.  In our case we do a squash-and-merge at the end so the final merge into master is cleaned up into a single commit and single nice description at that point, letting us keep the history with a series of commits during the review process.

derekbruening commented:
Merged!  Thank you again for the contribution.
